### PR TITLE
Start centrally handling "events"

### PR DESCRIPTION
**Pull request by bryanlatten**
1. Give a downstream implementer a funnel to "catch" all events as they flow through the system. With this, I could easily add an event system (such as Symfony EventDispatcher/Zend Event Manager) to be quickly snapped in by extending `AbstractCommand`
-- Long term could have pluggable systems for either
2. Allows `phpunit` to be run from source root as `./vendor/bin/phpunit`
3. Ensure that metrics/circuitbreakers are updated *before* events are "called" -- just incase the event solution goes off the rails :)

@lcf I'd love to get this running under TravisCI or an alternative, interested? 
I would also be very interested in undertaking a bit of legwork to decouple some of the dependencies (Zend Config/DI), which will allow easier integration in other projects -- especially when the dependencies themselves are updating much faster than before. For example, if we flattened out the configuration before creating a command, we could plug various configuration backends in at the factory level, and leave the command itself more generic. Next, I'd love to refit more of the requestCache/requestLog pieces to be optional, as well as driven by actual events instead of coded inline. What do you think?

### CODE REVIEW

**Comment by lcf** [comment]
Hi @bryanlatten. Thank you for you interest and pull requests! We wanted to get this running under Travis, to add a badge with tests status, just never got to it. What would you need for that?

Re decoupling from zend\di, that was discussed in https://github.com/odesk/phystrix/issues/4, what we've come is that decoupling for the sake of decoupling seems pointless. We may be missing something, if you have a clear idea - you can re-open that issue and let's discuss.

File: library/Odesk/Phystrix/AbstractCommand.php [hunk 1]
```diff
@@ -280,6 +279,23 @@ public function setServiceLocator(LocatorInterface $serviceLocator)
...
     }
 
+
+    /**
+     * Records list of actions that have taken place, with their exceptions
+     *
+     * @param string    $eventName  type from class constants EVENT_*
+     * @param Exception $e          exception that was thrown as part of $eventName
+     */
+    protected function handleEvent($eventName, \Exception $e = null )
```
Comment (lcf) [inline library/Odesk/Phystrix/AbstractCommand.php]: why does it accept the exception as well? we had it set and it was available already as public ->getExecutionException();

File: library/Odesk/Phystrix/AbstractCommand.php [hunk 2]
```diff
@@ -219,33 +219,32 @@ public function execute()
...
         } catch (BadRequestException $exception) {
             // Treated differently and allowed to propagate without any stats tracking or fallback logic
             $this->recordExecutionTime();
             throw $exception;
         } catch (Exception $exception) {
             $this->recordExecutionTime();
-            $this->executionException = $exception;
-            $this->executionEvents[] = self::EVENT_FAILURE;
             $metrics->markFailure();
+            $this->handleEvent(self::EVENT_FAILURE, $exception);
```
Comment (lcf) [inline library/Odesk/Phystrix/AbstractCommand.php]: to insure backwards compatibility and secure internal contract, we could consider calling a private method that will add the event to the list, so this cannot be overridden AND call a protected method, to allow for custom processing of events as they happen.

To think about it, you have access to the list of events already and the list is filled "atomically" with command execution. what is a use case, the need for processing these events as they happen, rather than after execution? @bryanlatten

File: library/Odesk/Phystrix/AbstractCommand.php (hunk 1 above)
Comment (bryanlatten) [inline library/Odesk/Phystrix/AbstractCommand.php]: @lcf  IMHO, the event and exception are actually tied together (one drove the other). This function allows for all event-related actions that need to take place (such as setting the exception, name) to be combined in a single location. In our downstream implementation, we then trigger an event based on all this information

File: library/Odesk/Phystrix/AbstractCommand.php (hunk 2 above)
Comment (bryanlatten) [inline library/Odesk/Phystrix/AbstractCommand.php]: @lcf I agree:
1. let's make handleEvent private, will serve as centralized function for performing the list addition, exception set (as per comment above), 
2. after, handleEvent can call either a protected function that can be overridden, or a registration-based closure-type assembly for downstream customization. 

#2 is super important for me, as phystrix does not have time-limiting or segfault-bypassing powers, so if the request times out, so does logging that it ever happened. We've deployed our fork (including all my PRs) to prod, and attached our symfony event system to use our logging system to record events as they happen. When requests time out completely, we can pinpoint where and how.

Comment (lcf) [inline library/Odesk/Phystrix/AbstractCommand.php]: > as phystrix does not (yet) have time-limiting or segfault-bypassing powers
You should configure a reasonable timeout for the underlying remote call (http connection timeout, db timeout etc) so that an exception is thrown when it happens.

Comment (bryanlatten) [inline library/Odesk/Phystrix/AbstractCommand.php]: @lcf of course I agree, but there is no guarantee that phystrix can enforce it. I have a simple solution in mind though. After a few more PR's though.

**Comment by bryanlatten** [comment]
@lcf ready for review: https://github.com/bryanlatten/phystrix/commit/66097ad5db3a029607030543fe6b9498473aaafa

File: library/Odesk/Phystrix/AbstractCommand.php [hunk 3]
```diff
@@ -280,6 +287,25 @@ public function setServiceLocator(LocatorInterface $serviceLocator)
...
     }
 
+
+    /**
+     * Logic to record events and exceptions as they take place
+     *
+     * @param string    $eventName  type from class constants EVENT_*
+     * @param Exception $e          exception that was thrown as part of $eventName
+     */
+    private function triggerEvent($eventName, \Exception $e = null )
```
Comment (lcf) [inline library/Odesk/Phystrix/AbstractCommand.php]: the conversation we started regarding this seem to be gone, probably due to the updated PR. Sorry, I never got to responding to your comment.

1)
> he event and exception are actually tied together (one drove the other).

I don't really see why. Event and execution exception are not directly related. There can only be one execution exception per command. The exception is available everywhere as ->getExecutionException();.

2) I don't really see much of a use case for this, but I don't mind either. If you could remove exception from this method and rename the private one to recordExecutionEvent and the protected one to processExecutionEvent - I will merge it

**Comment by bryanlatten** [comment]
@lcf great points all around. hopefully we can get this merged now: 667d59db1226f726f66c156c325012fa7e91ecc6

File: phpunit.xml.dist [hunk 4]
```diff
@@ -0,0 +1,36 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (lcf) [inline phpunit.xml.dist]: why was this file changed in this PR?

**Comment by lcf** [comment]
almost there!

@bryanlatten sorry if I wasn't clear about this

> If you could remove exception from this method and rename the private one to recordExecutionEvent and the protected one to processExecutionEvent - I will merge it

but seems like you have it the other way round. Phystrix code should be calling recordExecutionEvent which should be private, so that an execution event is recorded no matter what.

That private method should call protected processExecutionEvent, for extending classes to add functionality.

File: phpunit.xml.dist (hunk 4 above)
Comment (bryanlatten) [inline phpunit.xml.dist]: @lcf since this was my first PR, as stated in the notes, I added this file to provide the ability to run phpunit tests from the root

**Comment by bryanlatten** [comment]
@lcf yep. that reads way better, amended that commit
